.NET/Microsoft.Recognizers.Text.NumberWithUnit/Extractors/NumberWithUnitExtractor.cs

```diff
@@ -142,170 +159,150 @@ public List<ExtractResult> Extract(string source)
-                        mappingPrefix.Add(number.Start.Value, new PrefixUnitResult { Offset = offSet, UnitStr = unitStr });
-                    }
-                }
-            }
+                        if (maxlen != 0)
+                        {
+                            for (var i = 0; i < length + maxlen; i++)
+                            {
+                                matched[i + start] = true;
+                            }
 
-            foreach (var number in numbers)
-            {
-                if (number.Start == null || number.Length == null)
-                {
-                    continue;
-                }
+                            var substr = source.Substring(start, length + maxlen);
 
-                int start = (int)number.Start, length = (int)number.Length;
-                var maxFindLen = sourceLen - start - length;
+                            var er = new ExtractResult
+                            {
+                                Start = start,
+                                Length = length + maxlen,
+                                Text = substr,
+                                Type = this.config.ExtractType
+                            };
 
-                mappingPrefix.TryGetValue(start, out PrefixUnitResult prefixUnit);
+                            if (prefixUnit != null)
+                            {
+                                prefixMatched = true;
+                                er.Start -= prefixUnit.Offset;
+                                er.Length += prefixUnit.Offset;
+                                er.Text = prefixUnit.UnitStr + er.Text;
+                            }
 
-                if (maxFindLen > 0)
-                {
-                    var rightSub = source.Substring(start + length, maxFindLen);
-                    var unitMatch = suffixRegexes.Select(r => r.Matches(rightSub)).ToList();
+                            // Relative position will be used in Parser
+                            number.Start = start - er.Start;
+                            er.Data = number;
 
-                    var maxlen = 0;
-                    for (var i = 0; i < unitMatch.Count; i++)
-                    {
-                        foreach (Match m in unitMatch[i])
-                        {
-                            if (m.Length > 0)
+                            // Special treatment, handle cases like '2:00 pm', '00 pm' is not dimension
+                            var isNotUnit = false;
+                            if (er.Type.Equals(Constants.SYS_UNIT_DIMENSION))
                             {
-                                var endpos = m.Index + m.Length;
-                                if (m.Index >= 0)
-                                {
-                                    var midStr = rightSub.Substring(0, Math.Min(m.Index, rightSub.Length));
+                                var nonUnitMatch = this.config.PmNonUnitRegex.Matches(source);

```

Comment:
Can you use another variable to cache this results?
like: if (nonUnitMatch... == null){
    nonUnitMatch = this...
}